### PR TITLE
Grouped grid welford

### PR DESCRIPTION
**Pull request by naoyam**
Enables grouping of grid welford ops across iterations. Same functionality as the iteration grouping for `GridReduction`. This ins intended to improve the outer-norm grid persistence in batchnorm-like fusions.

To support the grouping, added `GroupedWelfordOp` and `kir::GroupedGridWelford`. In the case of grouping non-welford reductions, we support both grouping of multiple separate reduction expressions and grouping of the same expression across iterations. In the case of welford, I only enabled iteration grouping as I don't know if it's useful to be able to group multiple welford expressions. Specifically, `groupReductions()` is the function to group multiple `ReductionOp` exprs, which isn't extended to support welford. Iteration grouping is done with `IterationDomain::parallelize(ParallelType::Group)`.

There are a lot of mechanical changes as I added new classes `GroupedWelfordOp` and `kir::GroupedGridWelford`. Along the addition of the two classes, I modified some of the `WelfordOp` and `kir::GridWelford` code, which I think should be cleaner than before.

As the runtime implementation of the grouped grid welford, `ParallelReduce::welfordGroup` in fused_reduction.cu is added. It is mostly just a copy of `ParalleReduce::reduceGroup` but supports welford reductions. Thought about extending `reduceGroup` for welford as well, but that would lead to even more complicated template designs, which I'm not really comfortable.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 1]
```diff
@@ -258,78 +262,215 @@ class TORCH_CUDA_CU_API GroupedReductionOp : public Expr {
   bool is_allreduce_ = false;
 };
 
+//! Average, variance and N (count) vals for Welford
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:265]: Tried to clean the code up a little with a class to represent the three parameters of welford.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 2]
```diff
@@ -258,78 +262,215 @@ class TORCH_CUDA_CU_API GroupedReductionOp : public Expr {
...
       Val* out_N,
-      Val* init_avg,
-      Val* init_var,
-      Val* init_N,
       Val* in_avg,
       Val* in_var,
       Val* in_N,
+      Val* init_avg,
+      Val* init_var,
+      Val* init_N,
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:405]: Ordering of `out`, `in` and `init` seems to vary, which can be confusing and bug prone. Tried to make it always `out`, `in` and then `init` for consistency.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 3]
```diff
@@ -398,11 +398,12 @@ void IrPrinter::handle(const ReductionOp* rop) {
   indent() << "   = reduction( " << rop->in()
            << ", op = " << rop->getReductionOpType()
            << ", initial value = " << rop->init()
-           << ", allreduce = " << rop->isAllreduce() << " )\n";
+           << ", allreduce = " << (rop->isAllreduce() ? "true" : "false")
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:401]: All the changes in this file are mostly cleanup, except those newly added for `GroupedWelfordOp` and `GroupedGridWelford`. The cleanup is meant to make the printing more consistently done, whereas previously the KIR classes had been formatted rather differently.

**Comment by naoyam** [comment]
Code snippet from `FusionCrossIterationGroupedGridAllreduceWelford1`:

Without grouping:
```
  fused_reduction::ParallelReduce<1, 0, 3, 1, 0, 3, true, true> T2_reduction;
  #pragma unroll
  for(nvfuser_index_t i46 = 0; i46 < 2; ++i46) {
    int64_t i177;
    i177 = (((((nvfuser_index_t)blockIdx.x) * 32) + ((nvfuser_index_t)threadIdx.x)) * 2) + (i46 + nvfuser_zero);
    T2_reduction.reduce(
      RefTuple<float, float, int64_t>(T2[i46], T3[i46], T4[i46]),
      ConstRefTuple<float, float, int64_t>(T1[i46], 0, 1),
      VolatilePtrTuple<float, float, int64_t>(&T7[0], &T8[0], &T9[0]),
      &T10[0],
      PtrTuple<float, float, int64_t>(reinterpret_cast<float*>(shared_mem_avg), reinterpret_cast<float*>(shared_mem_var), reinterpret_cast<int64_t*>(shared_mem_n)),
      ((i65 < T0.size[0]) && (i177 < T0.size[1])),
      (i177 < T0.size[1]),
      LocalTuple<float, float, int64_t>(0.00000000000000000e+00, 0.00000000000000000e+00, 0),
      welfordCombine<float, int64_t>);
  }
```

With grouping:
```
  fused_reduction::ParallelReduce<1, 0, 3, 1, 0, 3, true, true> T2_reduction;
  T2_reduction.welfordGroup<2, float, int64_t>(RefTuple<float, float>(T2[0], T2[1]),
    RefTuple<float, float>(T3[0], T3[1]),
    RefTuple<int64_t, int64_t>(T4[0], T4[1]),
    ConstRefTuple<float, float>(T1[0], T1[1]),
    ConstRefTuple<float, float>(0, 0),
    ConstRefTuple<int64_t, int64_t>(1, 1),
    LocalTuple<float, float>(0.00000000000000000e+00, 0.00000000000000000e+00),
    LocalTuple<float, float>(0.00000000000000000e+00, 0.00000000000000000e+00),
    LocalTuple<int64_t, int64_t>(0, 0),
    VolatilePtrTuple<float, float>(&T7[0], &T7[((((nvfuser_index_t)gridDim.x) * ((nvfuser_index_t)gridDim.y)) * 32) * 1]),
    VolatilePtrTuple<float, float>(&T8[0], &T8[((((nvfuser_index_t)gridDim.x) * ((nvfuser_index_t)gridDim.y)) * 32) * 1]),
    VolatilePtrTuple<int64_t, int64_t>(&T9[0], &T9[((((nvfuser_index_t)gridDim.x) * ((nvfuser_index_t)gridDim.y)) * 32) * 1]),
    &T10[0],
    PtrTuple<float, float, int64_t>(reinterpret_cast<float*>(shared_mem_avg), reinterpret_cast<float*>(shared_mem_var), reinterpret_cast<int64_t*>(shared_mem_n)),
    LocalTuple<bool, bool>(((i65 < T0.size[0]) && (((((((nvfuser_index_t)blockIdx.x) * 32) + ((nvfuser_index_t)threadIdx.x)) * 2) + (0 + nvfuser_zero)) < T0.size[1])), ((i65 < T0.size[0]) && (((((((nvfuser_index_t)blockIdx.x) * 32) + ((nvfuser_index_t)threadIdx.x)) * 2) + (1 + nvfuser_zero)) < T0.size[1]))),
    LocalTuple<bool, bool>((((((((nvfuser_index_t)blockIdx.x) * 32) + ((nvfuser_index_t)threadIdx.x)) * 2) + (0 + nvfuser_zero)) < T0.size[1]), (((((((nvfuser_index_t)blockIdx.x) * 32) + ((nvfuser_index_t)threadIdx.x)) * 2) + (1 + nvfuser_zero)) < T0.size[1])));
```

**Comment by csarofeen** [comment]
> I don't know if it's useful to be able to group multiple welford expressions

I can't think of any cases off the top of my head.

File: torch/csrc/jit/codegen/cuda/codegen.cpp [hunk 4]
```diff
@@ -1872,6 +1882,154 @@ class CudaKernelGenerator : private OptOutConstDispatch {
...
+
+        data_types.arg(data_type);
+        index_types.arg(index_type);
+
+        auto work_buffer_offset = group_index == 0
+            ? "0"
+            : (genInline(grouped_gwop->buffer_stride()) + " * " +
+               std::to_string(group_index));
+
+        // Setup arguments fof avg, var, and N
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/codegen.cpp]: typo

File: torch/csrc/jit/codegen/cuda/grouped_reduction.h [hunk 5]
```diff
@@ -27,6 +27,10 @@ namespace cuda {
 //!   dimensions, the same transformations and the same axes to
 //!   reduce.
 //!
+//! Note that Welford is not allowed yet, though it should be
+//! technically straightforward to support horizontal fusions of
+//! welford ops. Unclear how common it would be in practie, though.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/grouped_reduction.h]: typo: practice

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 6]
```diff
@@ -649,138 +671,164 @@ void IrPrinter::handle(const kir::GridBroadcast* node) {
...
-    handle(welford_op->initAvg());
-    os_ << " initN=";
-    handle(welford_op->initN());
-  }
-  indent() << ", read_pred=";
+  indent() << welford_op->outAvg() << " (Avg),\n";
+  indent() << welford_op->outVar() << " (Var),\n";
+  indent() << welford_op->outN() << " (Count)\n";
+  indent() << " = Welford (\n";
+  indent() << kTab << welford_op->inAvg() << " (Avg),\n";
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: nit: seems you could `++indent_size_` here?

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 7]
```diff
@@ -649,138 +671,164 @@ void IrPrinter::handle(const kir::GridBroadcast* node) {
...
   } else {
     os_ << "nullptr";
   }
-  os_ << "\n";
+  os_ << ",\n";
+  indent() << "thread predicate = " << node->threadPredicate().toString()
+           << ",\n";
+  indent() << "allreduce = " << (welford_op->isAllreduce() ? "true" : "false")
+           << " )\n";
+  --indent_size_;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:793]: nit: should this be moved up a line, before the `)\n`? Hard to tell

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp [hunk 8]
```diff
@@ -2148,6 +2150,309 @@ TEST_F(NVFuserTest, FusionCrossIterationGroupedGridAllreduce4_CUDA) {
...
+  };
+
+  std::vector<Params> base_params;
+  base_params.push_back({256, 7, 7, 1, 8, 32, 2, 32, 4});
+  base_params.push_back({256, 7, 7, 1, 16, 16, 4, 50, 4});
+  base_params.push_back({128, 7, 7, 1, 16, 16, 4, 32, 4});
+  base_params.push_back({128, 14, 14, 1, 16, 16, 4, 32, 1});
+  base_params.push_back({128, 14, 14, 1, 16, 16, 2, 64, 2});
+  base_params.push_back({128, 14, 14, 1, 8, 32, 4, 50, 4});
+  base_params.push_back({128, 14, 14, 1, 8, 32, 2, 50, 4});
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:2450]: Are there any GPUs we frequently run tests on that this wouldn't fit on?

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp [hunk 9]
```diff
@@ -2148,6 +2150,309 @@ TEST_F(NVFuserTest, FusionCrossIterationGroupedGridAllreduce4_CUDA) {
...
+    auto tv6 = broadcast(tv3, bcast_pattern);
+    auto tv7 = broadcast(tv4, bcast_pattern);
+    auto tv8 = sub(tv1, tv5);
+    auto tv9 = add(tv8, tv6);
+    // auto tv10 = div(tv9, tv7);
+    // fusion.addOutput(tv10);
+    fusion.addOutput(tv9);
+
+    // Schedule the fusion as it will be done by the persistent
+    // scheduler
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:2322]: Didn't check this schedule carefully, I assume once outer grid persistence is in we'll just send it to the segmenter/schedulers.

**Review by csarofeen — APPROVED** [review]
I didn't see any obvious errors. LGTM

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 7 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp:793]: I guess I just accidentally added two decrement lines. Removed one.

File: torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp (hunk 8 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/test/test_gpu_fused_reduction.cpp:2450]: Forgot to add a check. Thanks for the comment.

**Pull request merged**